flame/data.py

```
@@ -381,57 +380,134 @@
+                    cu_seqlens = [torch.tensor([0], device=batch['input_ids'].device)] # Match device
+                    # Find positions *after* EOS tokens
+                    eos_positions = torch.where(batch['input_ids'].eq(self.tokenizer.eos_token_id))[1] + 1
+                    # Ensure eos_positions is on the correct device if empty
+                    if eos_positions.numel() > 0:
+                         cu_seqlens.append(eos_positions)
+                    # Handle case where the sequence doesn't end with EOS
                     if batch['input_ids'][0, -1] != self.tokenizer.eos_token_id:
-                        cu_seqlens.append(torch.tensor([len(batch['input_ids'][0])]))
-                    batch['cu_seqlens'] = torch.cat(cu_seqlens, dim=0).to(dtype=torch.int32)
+                         # Only add the final length if the last found EOS wasn't already the end
+                        if eos_positions.numel() == 0 or eos_positions[-1] != batch['input_ids'].size(1):
```

Ruff (0.8.2)

Use a single `if` statement instead of nested `if` statements

(SIM102)